_source/_posts/2021-12-18-native-java-helidon.adoc

```diff
@@ -0,0 +1,403 @@
+
+**Prerequisites**
+
+- https://sdkman.io/[SDKMAN]
+- https://maven.apache.org[Maven]
+- https://httpie.io/[HTTPie] (a better version of cURL)
+- An https://developer.okta.com[Okta Developer] Account
+
+toc::[]
+
+== Get Started with Helidon
+
+I created a GitHub repository you can clone and run to get started with Helidon quickly.
+
+[source,shell]
+----
+git clone https://github.com/oktadev/native-java-examples.git
+cd native-java-examples/helidon
+----
+
+I'll show you how I created this example in a section below. First, let's see how to run it.
+
+Open the `helidon` directory in your favorite IDE, so you have easy access to the example's project files.
+
+TIP: If you just want to see how to build a native image, skip to the link:#build-native-images-for-helidon[build native images for Helidon] section.
+
+== Install a JDK with GraalVM
+
+You will need a JDK with GraalVM and its `native-image` compiler. Using SDKMAN, run the following command and set it as the default:
+
+[source,shell]
+----
+sdk install java 21.3.0.r17-grl
+----
+
+Add the native extension to the JDK:
+
+[source,shell]
+----
+gu install native-image
+----
+
+== Run a Helidon Java API
+
+In a terminal window, run `mvn package` to package the app into a JAR. Then use `java -jar` to run it.
+
+[source,shell]
+----
+mvn package
+java -jar target/helidon.jar
+----
+
+TIP: Helidon's Maven Plugin doesn't have a _run_ command. You can install the https://helidon.io/docs/v2/#/about/05_cli[Helidon CLI] and run `helidon dev`. Running this command allows you to change code and automatically recompile + restart your application.
+
+If you open another terminal window and try to access it with HTTPie, you'll get a 401 Unauthorized error.
+
+[source,shell]
+----
+$ http :8080/hello
+
+HTTP/1.1 401 Unauthorized
+Content-Length: 0
+Date: Fri, 3 Dec 2021 06:11:25 -0700
+connection: keep-alive
+----
+
+To make it so you can access this endpoint, you'll need to generate an OAuth 2.0 access token and update the JWKS (JSON Web Key Sets) URL to yours (in this project's `application.yml`).
+
+NOTE: If you're unsure what OIDC and OAuth 2.0 are, see link:/blog/2017/06/21/what-the-heck-is-oauth[What the Heck is OAuth?].
+
+{% include setup/cli.md type="spa" loginRedirectUri="https://oidcdebugger.com/debug" logoutRedirectUri="https://oidcdebugger.com" %}
+
+Take note of the `clientId` and `issuer` values. You'll need those to get an access token and to configure each framework for JWT authentication.
+
+Open `src/main/resources/META-INF/microprofile-config.properties` and change the Okta URL to match yours.
+
+[source,properties]
+----
+mp.jwt.verify.publickey.location=https://{yourOktaDomain}/oauth2/default/v1/keys
+mp.jwt.verify.issuer=https://{yourOktaDomain}/oauth2/default
```

Comment:
nit: `https://{yourOktaDomain}/oauth2/default` should probably be {issuer} in all places (as that is what the CLI outputs)

Reply:
That said, we use `yourOktaDomain` all over the place, so that is fairly consistent too, so feel free to ignore this